### ISSUE TITLE
notifying enumerators/data collectors of missing answers without using `required`

### ISSUE THREAD

**Issue by ernestguevarra** [ecohealthalliance/xlsform#4]
@mekaczmarek asked whether and how we can notify enumerators/data collectors of missing answers without using `required` field for the questions.

Overview response is **yes, this is possible**: 

* using `calculate` field to calculate whether any of the questions asked have no answers (value is null);
* if the result of the calculation indicates that at least one question has a value null, a `note` field can be created to notify the enumerator/data collector that one or more of the earlier questions have missing answers and for the enumerator/data collector to go back and review accordingly (this is a soft constraint); or,
* if the result of the calculation indicates that at least one question has a value null, an `acknowledge` field can be created and specified as `required` but also `read_only` to notify the enumerator/data collector that one ore more of the earlier questions have missing answers and for the enumerator/data collector to go back and review accordingly (this is a hard constraint because the enumerator/data collector can only progress if he/she acknowledges the acknowledge field but since it is read-only, then the only way to progress is to provide answers to all the questions).

A soft constraint may be more appropriate and acceptable to the hard constraint depending on the project and the context.

The specific approaches to address this are demonstrated in the following XLSForm - https://docs.google.com/spreadsheets/d/1A9QQwukZUfiAxqnOfLENldwPfMYHysNclfCVklYd1GE/edit?usp=sharing

Following are examples on how to implement the solutions above in the XLSForm specifications:

1. Using `calculate` field with hard constraint

type     | name     | label     | appearance     | constraint     | relevant     | calculation     | read_only     | required     
:--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :---  
begin group | set1 | Using calculate - hard constraint | field-list | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; 
integer | q1 | Question 1 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; 
integer | q2 | Question 2 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp;
calculate | c1 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | if(${q1} = null or ${q2} = null, 1, 0) | &nbsp; | &nbsp;
acknowledge | check1 | You have left one or more of the questions in this set/module unanswered. Please review and complete all questions accordingly. | &nbsp; | ${c1} = null | ${c1} = null | &nbsp; | TRUE | TRUE

2. Without using `calculate` field with hard constraint

type     | name     | label     | appearance     | constraint     | relevant     | calculation     | read_only     | required     
:--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :---  
begin group | set1 | Without using calculate - hard constraint | field-list | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; 
integer | q1 | Question 1 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; 
integer | q2 | Question 2 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp;
acknowledge | check1 | You have left one or more of the questions in this set/module unanswered. Please review and complete all questions accordingly. | &nbsp; | ${q1} = null or ${q2} = null | ${q1} = null or ${q2} = null | &nbsp; | TRUE | TRUE

3. Using `calculate` field with soft constraint

type     | name     | label     | appearance     | constraint     | relevant     | calculation     | read_only     | required     
:--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :---  
begin group | set1 | Using calculate - soft constraint | field-list | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; 
integer | q1 | Question 1 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; 
integer | q2 | Question 2 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp;
calculate | c1 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | if(${q1} = null or ${q2} = null, 1, 0) | &nbsp; | &nbsp;
note | note1 | You have left one or more of the questions in this set/module unanswered. Please review and complete all questions accordingly. | &nbsp; | ${c1} = null | ${c1} = null | &nbsp; | TRUE | &nbsp;

4. Without using `calculate` field with soft constraint

type     | name     | label     | appearance     | constraint     | relevant     | calculation     | read_only     | required     
:--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :---  
begin group | set1 | Without using calculate - hard constraint | field-list | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; 
integer | q1 | Question 1 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; 
integer | q2 | Question 2 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp;
acknowledge | note1 | You have left one or more of the questions in this set/module unanswered. Please review and complete all questions accordingly. | &nbsp; | ${q1} = null or ${q2} = null | ${q1} = null or ${q2} = null | &nbsp; | TRUE | &nbsp;

**Comment by mekaczmarek** [comment]
Thanks @ernestguevarra ! 
I was able to figure this out, and for a non-integer example was using an empty string '' as my "null." Though I am a bit confused about the first example shown here - the calculate is based on returning a "null" but if it is null it will be assigned a "1", correct?

**Comment by mekaczmarek** [comment]
There are a number of questions with different skip logic in my form - which means some of them wouldn't necessarily be answered unless a specific previous answer was picked. Is there any way to write a calculate that could handle dependencies? This may be overly complicated and unnecessary, but I thought I should ask before I gave up.

**Comment by ernestguevarra** [comment]
Hi! This can be done. Depending on how your skip logic is, replicate the dependencies using `if()` functions; this can get very complex. There is also a way to calculate a sequence of answers/values and then check for null by indexing. Will try to get an example going before we talk. Don't give up yet. This is doable I think!

**Comment by ernestguevarra** [comment]
> Thanks @ernestguevarra ! 
> 
> I was able to figure this out, and for a non-integer example was using an empty string '' as my "null." Though I am a bit confused about the first example shown here - the calculate is based on returning a "null" but if it is null it will be assigned a "1", correct? 

Yes, not really needed but just wanted to show the calculate step. The recoding can be the other way, it is really just to distinguish whether there are nulls
